scripts/prepare-subcharts

```diff
@@ -0,0 +1,125 @@
+	if [[ ${repo} == file://* ]]; then
+		charts_path=${repo/file:\/\//${f}/charts/}
+		if [[ -d ${charts_path} ]]; then
+			# Chart already exists
+			((i=i+1))
+			continue
+		fi
+		if [[ -f ${charts_path%/*}/package.yaml ]]; then
+			# Prepare the subchart and pull it in
+			subchart=$(basename -- ${charts_path%/*})
+			mirrored_dependencies+=("${subchart}")
+			prepare_subchart () {
+				local CURRENT_CHART=${CHART}
+				local CURRENT_INDEX=${i}
+				CHART=${subchart} ./scripts/prepare
+				CHART=${CURRENT_CHART}
+				i=${CURRENT_INDEX}
+			}
+			prepare_subchart
+		fi
+		overlay_path=${repo/file:\/\//${f}/overlay/}
+		if [[ -d $overlay_path ]]; then
+			# Pull in the subchart from the overlay/ directory
+			cp -R $overlay_path $charts_path
+			overlaid_dependencies+=("${charts_path}")
+		fi
+	fi
+	((i=i+1))
+done
+
+# Update the current requirements.lock and requirements.yaml to keep track of the overlay/ dependencies
+pwd=$(pwd)
+cd ${f}/charts
+helm dependency update
+cd $pwd
+
+# Remove any dependencies that came from the overlay/ directory
+for dep in "${overlaid_dependencies[@]}"; do
+	rm -rf $dep
+done
+
+# Cleanup any dependencies that came from mirrored package.yamls
+for subchart in "${mirrored_dependencies[@]}"; do
+	cleanup_subchart () {
+		local CURRENT_CHART=${CHART}
+		CHART=${subchart} ./scripts/clean
+		CHART=${CURRENT_CHART}
+	}
+	cleanup_subchart
+done
+
+if ! [[ -d ${f}/charts/charts ]]; then
+	exit 0
+fi
+
+# If it is possible to patch a subchart, create a local chart archive for it
+# Otherwise, it should only be stored as a tgz file to prevent users from accidentally patching it
+i=0
+while ! [[ -z $(yq r ${f}/charts/requirements.yaml "dependencies[${i}]") ]]; do
+	name=$(yq r ${f}/charts/requirements.yaml "dependencies[${i}].name")
+	version=$(yq r ${f}/charts/requirements.yaml "dependencies[${i}].version")
+	charts_path=${f}/charts/charts/${name}
+	subchart_tgz=${name}-${version}.tgz
+	repo=$(yq r ${f}/charts/requirements.yaml "dependencies[${i}].repository")
+	if [[ ${repo} == file://* ]]; then
+		# Subcharts that are tracked locally should be directly edited, not via a patch
+		if [[ -f ${subchart_tgz} ]]; then
+			rm -rf charts_path
+		fi
+		((i=i+1))
+		continue
+	fi
+	subchart_path=$(ls ${f}/charts/charts/${subchart_tgz})
+	if [[ ${version} == *'*'* ]]; then
+		# Subcharts that are tracked remotely must have a fixed version number
+		if [[ -f ${subchart_path} ]]; then
+			rm -rf ${charts_path}
+		fi
+		((i=i+1))
+		continue
+	fi
+	# Subchart could be patched
+	if [[ -f ${subchart_path} ]]; then
+		tar xvzf ${f}/charts/charts/${subchart_tgz} -C ${f}/charts/charts > /dev/null 2>&1
+		rm ${f}/charts/charts/${subchart_tgz}
+	fi
+	((i=i+1))
+done
+
+# Undo the patch on the requirements.yaml
+if [[ -f ${f}/requirements.patch ]] && [[ -s ${f}/requirements.patch ]]; then
+	patch -R -E -p3 -d ${f}/charts < ${f}/requirements.patch
+	rm ${f}/requirements.patch
+fi
```

Comment:
Everything in the sub-chart folder already gets cleaned up by `make clean` right ?

Reply:
Yes, but this isn't located in the subchart folder; this is the extracted patch specifically created from the awk statement (explained in https://github.com/rancher/charts/pull/579#discussion_r481832652) in order to support patches on added dependencies as described in https://github.com/rancher/charts/pull/579#discussion_r481825133.

Since this is created by the prepare-subcharts scripts, it's also deleted by the prepare-subcharts script.

Reply:
If we need to edit the main chart, and requirements, and a sub-chart, all those patches would just exist in the primary patch file right ? This is just temporary.

Reply:
Yeah, all of those would be in the same patch file. This just temporarily strips the requirements.yaml changes out.